File: LeetCode/backtracking/47.permutations-ii.py
#
# @lc app=leetcode id=47 lang=python3
#
# [47] Permutations II
#

# @lc code=start
class Solution:
    result = []
    nums =[]
    used = []
    def backtrack(self, partialSol):
        if len(partialSol) == len(self.nums):
            self.result.append(partialSol.copy())

        for (idx, num) in enumerate(self.nums):
            if self.used[idx]:
                continue
            if (idx > 0 and self.nums[idx] == self.nums[idx-1] and (not self.used[idx-1])): # 这里是关键 (not self.used[idx-1])
                continue
            self.used[idx] = True
            partialSol.append(num)
            self.backtrack(partialSol)
            self.used[idx] = False
            partialSol.pop(-1)
        
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        self.result = []
        self.nums = nums
        self.nums.sort()
        self.used = [False for i in range(len(nums))]
        self.backtrack([])
        return self.result
        
# @lc code=end

# https://labuladong.github.io/algo/1/9/#排列元素可重不可复选

# 另一剪法 
class Solution:
    result = []
    nums =[]
    used = []
    def backtrack(self, partialSol):
        if len(partialSol) == len(self.nums):
            self.result.append(partialSol.copy())

        prevNum = -100  # 记录同一层左边sibling元素的值
        for (idx, num) in enumerate(self.nums):
            if self.used[idx]:
                continue
            if (self.nums[idx] == prevNum): 
                continue
            self.used[idx] = True
            partialSol.append(num)
            prevNum = num
            self.backtrack(partialSol)
            self.used[idx] = False
            partialSol.pop(-1)
        
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        self.result = []
        self.nums = nums
        self.nums.sort()
        self.used = [False for i in range(len(nums))]
        self.backtrack([])
        return self.result

File: LeetCode_Go/Backtrack/47.permutations-ii.go
/*
 * @lc app=leetcode id=47 lang=golang
 *
 * [47] Permutations II
 */

// @lc code=start

// Given a collection of numbers, nums, that might contain duplicates, return all possible unique permutations in any order.

// two factors prune:  1. (ref.46.permutations) use usedNums, []bool as the alternative to choiceSet, skip numbers that has been used. 2. sort the input and skip current choice num[i] if same number has been visited in current call stack iteration.

func permuteUnique(nums []int) [][]int {
	slices.Sort(nums)
	ans := [][]int{}
	used := make([]bool, len(nums))
	backtracking(nums, []int{}, used, &ans)
	return ans
}

func backtracking(nums, perm []int, used []bool, ans *[][]int) {
	// exit: when the length is correct
	if len(perm) == len(nums) {
		// save to ans
		temp := make([]int, len(perm))
		copy(temp, perm)
		*ans = append(*ans, temp)
		return
	}

	// reinitialize lastAdd in every call stack, to skip branches that have already been visited
	lastVisited := -11 // init value must not appear in nums
	for i, num := range nums {
		// prune 1: used, prune 2: deal with duplicate
		if used[i] || (nums[i] == lastVisited) {
			continue
		}
		// make choice
		used[i] = true
		perm = append(perm, num)
		lastVisited = num
		backtracking(nums, perm, used, ans)
		// undo choice
		used[i] = false
		perm = perm[:len(perm)-1]
	}
}

// @lc code=end

